add mask_convert_to_lp to support bool->fp16/bf16 convert

zhuhaozhe opened the pull request:
Fix
https://github.com/pytorch/pytorch/issues/117624
https://github.com/pytorch/pytorch/issues/117627

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #117830



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

zhuhaozhe committed:
add mask_convert_to_lp to support bool->fp16/bf16 convert

[ghstack-poisoned]

zhuhaozhe committed:
Update on "add mask_convert_to_lp to support bool->fp16/bf16 convert"

Fix
https://github.com/pytorch/pytorch/issues/117624
https://github.com/pytorch/pytorch/issues/117627




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

jgong5 commented on torch/_inductor/codegen/cpp_prefix.h:
```diff
@@ -364,6 +364,12 @@ inline at::vec::Vectorized<float> mask_convert_to_float(at::vec::Vectorized<floa
   return at::vec::Vectorized<float>::blendv(zeros, ones, src);
 }
 
+template <typename scalar_t>
+inline at::vec::Vectorized<scalar_t> mask_convert_to_lp(at::vec::Vectorized<float> src) {
```
```suggestion
template <typename scalar_t>
inline
typename std::enable_if<std::is_same<scalar_t, bfloat16>::value || std::is_same<scalar_t, half>::value, at::vec::Vectorized<scalar_t>>::type
mask_convert_to_lowp(at::vec::Vectorized<float> src) {
```

jgong5 commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -1345,6 +1345,8 @@ def to_dtype(x, dtype, src_dtype=None):
         if opt_ctx_x.dtype == torch.bool and dtype in (torch.float, torch.float32):
             return f"mask_convert_to_float({x})"
+        if opt_ctx_x.dtype == torch.bool and dtype in DTYPE_LOWP_FP:
+            return f"mask_convert_to_lp<{DTYPE_TO_CPP[dtype]}>({x})"
```
```suggestion
            return f"mask_convert_to_lowp<{DTYPE_TO_CPP[dtype]}>({x})"
```

jgong5 commented on test/inductor/test_cpu_repro.py:
```diff
@@ -2765,6 +2765,22 @@ def forward(self, x):
                 "Vectorized<float>::loadu(tmpbuf.data())", 0, exactly=True
             ).run(code)
 
+    def test_logical_op_store_to_lp_data_dtype(self):
```
```suggestion
    def test_logical_op_store_to_lowp_data_dtype(self):
```

zhuhaozhe replied:
thanks for advice, all revised.

zhuhaozhe committed:
Update on "add mask_convert_to_lp to support bool->fp16/bf16 convert"

Fix
https://github.com/pytorch/pytorch/issues/117624
https://github.com/pytorch/pytorch/issues/117627




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

zhuhaozhe commented:
@pytorchbot  rebase

zhuhaozhe committed:
Update on "add mask_convert_to_lp to support bool->fp16/bf16 convert"

Fix
https://github.com/pytorch/pytorch/issues/117624
https://github.com/pytorch/pytorch/issues/117627




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

zhuhaozhe commented:
@pytorchbot merge

zhuhaozhe commented:
@pytorchbot merge

zhuhaozhe commented:
@pytorchbot merge

zhuhaozhe committed:
Update on "add mask_convert_to_lp to support bool->fp16/bf16 convert"

Fix
https://github.com/pytorch/pytorch/issues/117624
https://github.com/pytorch/pytorch/issues/117627




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]
